### PR TITLE
Archiving of Hubs

### PR DESCRIPTION
**Pull request by techsmyth**
Extend Hub with a field to control its visibility, which can be ACTIVE, ARCHIVED or DEMO. 
- demo visibility is for later

api:
* New mutation to allow updating the hub visibility: updateHubVisibility
* hubs() query: takes optional argument to allow filtering of hubs by visiblity; defaults to active hubs
* metadata query metrics are updated to show only active hubs, challenges, opportunities counts

modules:
* new HubFilter module to manage the filtering of Hubs based on visibility in a consistent way

samples:
* new sample query to get hubs by visibility
* new sample query to get roles by visibility of Hubs for which credentials are for

Key changes:
* Updated HubAuthorizationService so that if Hub is archived then the HubAdmin does not get assigned privileges, nor do any privileges cascade down into the contained entities. 

Note: due to VS Code changes had to mark hubID on Challenge / Opportunity to be optional as otherwise type validations were failing.

Search:
* search is filtering out results based on users access, so to test it out in action run the search as a **non-global-admin** user

There are likely to be additional enhancements identified re filtering as we move demo onto prod + do the FFE e.g.
* metadata metrics to be specified to the url being used?
* ...

To look at it the authorization in action, run the following query:
query {
  hubs {
    nameID
    visibility
    authorization {
      anonymousReadAccess
      credentialRules {
        grantedPrivileges
        resourceID
        type
      }
    }
  }
}

and then run the mutation to update the visibility to be archived. When in active visibility then credential rules looks like: 
image

when archived:

image

### CODE REVIEW

File: graphql-samples/mutations/update/update-hub-visibility [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+  updateHubVisibility(visibilityData: $visibilityData) {
+    id
+  }
+}
+
+Variables:
+{
+  "visibilityData": {
+    "hubID": "uuid_nameid",
+    "visibility": "active"
```
Comment (valentinyanakiev) [inline graphql-samples/mutations/update/update-hub-visibility]:
```suggestion
    "visibility": "ACTIVE"
```

Comment (valentinyanakiev) [reply graphql-samples/mutations/update/update-hub-visibility]: `active`, lowercase, is invalid value

File: src/domain/challenge/challenge/challenge.entity.ts [hunk 2]
```diff
@@ -42,7 +42,7 @@ export class Challenge extends BaseChallenge implements IChallenge {
   parentHub?: Hub;
 
   @Column()
-  hubID!: string;
+  hubID?: string;
```
Comment (valentinyanakiev) [inline src/domain/challenge/challenge/challenge.entity.ts]: Currently, `hubID` is mandatory. We don't really support nested `Challenge`. Why was this changed?

File: src/domain/challenge/hub/hub.service.ts [hunk 3]
```diff
@@ -256,7 +278,14 @@ export class HubService {
       .leftJoinAndSelect('challenge.opportunities', 'opportunities')
       .getMany();
 
-    const sortedHubs = hubsDataForSorting.sort((a, b) => {
+    const visibleHubs = hubsDataForSorting.filter(hub => {
```
Comment (valentinyanakiev) [inline src/domain/challenge/hub/hub.service.ts:281]: You could do the filter and sort in the query itself. It will be much faster and the code will be less.

**Review by valentinyanakiev — CHANGES_REQUESTED** [review]

File: graphql-samples/mutations/update/update-hub-visibility (hunk 1 above)
Comment (techsmyth) [reply graphql-samples/mutations/update/update-hub-visibility]: Fixed

File: src/domain/challenge/challenge/challenge.entity.ts (hunk 2 above)
Comment (techsmyth) [reply src/domain/challenge/challenge/challenge.entity.ts]: There has been a change in VS Code which means that it no longer considers the repositories between Hub / Challenge / Opportunity the same as BaseChallenge if there are mandatory fields in the entity definition. So the whole project is flashing with errors without this change.

File: src/domain/challenge/hub/hub.service.ts (hunk 3 above)
Comment (techsmyth) [reply src/domain/challenge/hub/hub.service.ts:281]: Yes of course that is faster. But I also wanted to go through the same code path (in the hubFilter service) and also not spend time working a more complex db query. If in the future the number of hubs gets into the 10,000 range then worth considering optimising this, but could not justify that now.

**Review by valentinyanakiev — APPROVED** [review]
